### ISSUE TITLE
Scaling issues on Mac

### ISSUE THREAD

**Issue by pontusm** [buggins/dlangui#529]
There seems to be a scaling issue with UIs on MacOS. On my Macbook Air I get very large UIs compared to when running on Windows. I have included a screenshot below of the "example1" demo app but I'm not sure if it's clear what is happening here.

![screen shot 2017-12-04 at 20 49 33](https://user-images.githubusercontent.com/144051/33573207-7357038a-d935-11e7-82e4-91bc0157a67b.png)

I'm kind of expecting the UI to be the exact same size when comparing PC and Mac on the same monitor. Or am I missing something here?

**Comment by pontusm** [comment]
For comparison, here is the exact same example app running on Windows.

![image](https://user-images.githubusercontent.com/144051/33573795-3ad4868e-d937-11e7-9f7b-842ba39a89ef.png)

**Comment by pontusm** [comment]
Additionally, the difference is quite a bit more exaggerated when comparing DLangIDE between Mac and Windows. Both of these screenshots are from the same monitor, with the app maximized to fill the screen.

Mac:
![screen shot 2017-12-04 at 21 12 01](https://user-images.githubusercontent.com/144051/33574081-2e228e94-d938-11e7-8d9e-ef4da7018a8d.png)

PC:
![image](https://user-images.githubusercontent.com/144051/33574093-365a6dd4-d938-11e7-81e0-fdf8dc09174a.png)

**Comment by pontusm** [comment]
Done some more research around this and learned a lot :)

I created a little test app with a window of size 300x300 and put three 100x100 images in it side by side. I expected the images to fill up the window.

On my MacBook it looks perfect because the size is exactly as I specified, i.e. 300x300 pixels. On the PC the window is scaled up to 400x400 pixels and this introduces 100 pixels of empty space on the side.

One could argue that this is correct. On Windows 10 the default DPI is 96, which is also used by dlangui. The pointsToPixels method in types.d will scale the original size of the window from 300x300 to 400x400 because the standard is 72 points per inch. However, as a developer making apps for both Mac and PC the behavior is a bit unexpected. It makes it harder to write nice looking programs imho.

I would either expect the images to be scaled up too, so that each image is scaled to the right dpi. This would give me the expected result with the images filling the window on both platforms.

Or, I would expect the app window to be 300x300 pixels also on PC as long as the DPI is not too high. When the DPI goes over a certain threshold, it would switch to a higher res, and scale up. This would probably look better than scaling the images, and it would only require images at certain predetermined resolutions. This is similar to how image resources work on iOS.

Now, I'm pretty new to this lib and have only tried it out for a couple of days so I may be missing something here. Maybe the resource system is supposed to handle this automatically, and select the appropriate image size. But the fact remains, that having images for every thinkable dpi is a bit crazy.

I'm not very familiar with Android development, but I did some reading and I think this is how the "density" property of DisplayMetrics on Android works. It is not the exact dpi, but rather the closest step that makes sense to scale the UI to. https://developer.android.com/reference/android/util/DisplayMetrics.html

**Comment by mermoldy** [comment]
An example app on 15.4-inch display (2880-by-1800 resolution):
<img width="1245" alt="screenshot 2017-12-17 03 04 19" src="https://user-images.githubusercontent.com/6282851/34075741-db3bd49c-e2d8-11e7-984b-7cb9dddae6ff.png">
<img width="234" alt="screenshot 2017-12-17 03 04 48" src="https://user-images.githubusercontent.com/6282851/34075743-e2cdfd7a-e2d8-11e7-960f-140d1283b1c6.png">

**Comment by andresclari** [comment]
Yeah, on retina displays seems scaling is not working, which makes it pretty unusable.

**Comment by and3md** [comment]
Does setting DPI with `overrideScreenDPI` improve the situation?

**Comment by andresclari** [comment]
Tried it out, unfortunately no. Some elements respond to it, mostly text tho, but overall looks pretty bad.

![baddpi](https://user-images.githubusercontent.com/74197/35429682-6bd8c60c-0275-11e8-9ba7-a253c4826c48.png)

This is adding "overrideScreenDPI(200)" also the platform code should be detecting the actual screen DPI from the system, and follow the proper scaling setup for the display, since that's bound to change the actual rendering DPI.

Controls like scrollbars don't seem to scale at all. The hdpi images don't seem to get used either, the toolbar for instance is tiny.

**Comment by and3md** [comment]
Maybe it's a graphics problem. I noticed that the default template has no hdpi scrollbar graphics. The icons are probably the same.

**Comment by and3md** [comment]
@andresclari In v0.9.174 version scrollbars should be displayed correctly.

>he platform code should be detecting the actual screen DPI from the system, and follow the proper scaling setup for the display, since that's bound to change the actual rendering DPI.

Yes you are right! Is your testing machine with one display or more?

**Comment by andresclari** [comment]
Scrollbars are indeed fixed 💯

I'm running 2 displays, main is 4k set to scale to look like 144p, secondary screen is 1080p normal display.

Started a new project, figure with some luck by the time I finished we might have solved the scaling issues on macOS. On Windows the app behaves properly tho.

For now I'm gonna use the "overrideScreenDPI(192)" on the Mac, however, if I move the window to the 1080p display, obviously everything looks huge. Duh.

From the logs:
```
sdlUpdateScreenDpi: systemScreenDPI=107                                           // 1080p screen
sdlUpdateScreenDpi: systemScreenDPI is overrided = 192                    // 4k screen
sdlUpdateScreenDpi: systemScreenDPI is changed from 192 to 107    // 4k back to 1080p dpi?
```

So, looks like SDL detects the correct DPI initially, but then gets the DPI for the 1080p and sets it to that.

We'd need to get the DPI for the current screen where the window is, and use that DPI instead. And listen to the window manager events when changing screens to switch the DPI accordingly to what's the display the window is being drawn in.

I'd say each window should keep track of it's current DPI and when the platform receives a new DPI when changing screens, it should perform a layout update to redraw in the proper DPI (unless override DPI is set I suppose).

**Comment by and3md** [comment]
Currently dlangui gets dpi always from first display and its stored in types.d file. I think it should be in Window class.  That's need to be changed. 

The log says only that system dpi is changed, but is still overriden. Maybe that need clarification too.

**Comment by joelcnz** [comment]
Print is still tiny, on macOS Mojave with Retina display. (I was going to post a snapshot, but then realized that wouldn't work).

**Comment by GrimMaple** [comment]
@buggins is it at all possible that this issue is fixed in your recent work?
